### PR TITLE
Add EmGithub liquid tag

### LINKED ISSUE #10151: Support emgithub.com embed using Liquid Tags
In my blog posts I want to be able to embed the github repo code directly using [emgithub.com](https://emgithub.com/) 
This way there is no need to create a separate gist file as you can just target any lines on existing github repo. 

image

Here is the [link example](https://emgithub.com/embed.js?target=https%3A%2F%2Fgithub.com%2Fkuncevic%2Fprogressive-state-management-with-ngxs%2Fblob%2Fmaster%2Fsrc%2Fapp%2Freactive-service%2Fservices%2Fbasic-example%2Fcounter.service.ts%23L16-L40&style=zenburn&showBorder=on&showLineNumbers=on&showFileMeta=on) generated by emgithub. Currently there is no way to embed [emgithub.com](https://emgithub.com/)  using Liquid Tags.

Having [emgithub.com](https://emgithub.com/)  support would reduce a lot of hassle of providing code-samples within the [dev.to](dev.to) posts as well as maintaining the samples if they needs to be updated in the future.

### PR DESCRIPTION
**Pull request by ShubhankarKG**
## What type of PR is this? (check all applicable)

- [x] Feature

## Description
This is a preliminary support for emgithub tag, there's a slight bit of problem and I think it best to have a discussion on this before proceeding further.

## Related Tickets & Documents
Fixes #10151 

## QA Instructions, Screenshots, Recordings

1. Add a new post with the liquid tag **emgithub** and the Github link. for eg: 
```liquid
{% emgithub https://github.com/CommunityOfCoders/COCWebsite/blob/master/new_client/package.json#L2-L7 %}
```
Screenshot from 2020-11-05 09-59-15

2. Go to preview tab.
Screenshot from 2020-11-05 09-59-25

3. Publish post and see result.
Screenshot from 2020-11-05 09-59-48

## Added tests?

- [x] No, and this is why: Will write once the flow gets finalized.

## Added to documentation?
- [ ] Docs.forem.com
- [ ] README
- [ ] No documentation needed

### CODE REVIEW

File: app/liquid_tags/emgithub_tag.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
+require "cgi"
+
+class EmgithubTag < LiquidTagBase
+  PARTIAL = "liquids/emgithub".freeze
+  OPTION_REGEXP = /\A(showBorder|showLineNumbers|showFileMeta)*\z/.freeze
+  LINK_REGEXP = %r{\A(http|https)://github.com/[a-zA-Z0-9]*/[a-zA-Z0-9]*/blob/[a-zA-Z0-9_./#-]*\z}.freeze
```
Comment (rhymes) [inline app/liquid_tags/emgithub_tag.rb:6]: Mmm this regexp would also match something like `https://github.com///blob/`

It's probably easier to just check if the URL starts with github.com, it's followed by `/\w+/\w+/blob/` and the optional chars you put at the end. Remember that `*` means 0 ore more, we need at least 1 per each part of the path. You can test on https://rubular.com/

File: app/liquid_tags/emgithub_tag.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    validated_options = options.map { |option| valid_option(option) }.reject(&:nil?)
+    raise StandardError, "Invalid Options" unless options.empty? || !validated_options.empty?
+
+    options.length.zero? ? "" : validated_options.join("=on&")
+  end
+
+  def parse_link(link)
+    stripped_link = ActionController::Base.helpers.strip_tags(link)
+    the_link = stripped_link.split(" ").first
+    raise StandardError, "Invalid Github URL" unless valid_link?(the_link)
```
Comment (rhymes) [inline app/liquid_tags/emgithub_tag.rb:45]: this should be `GitHub`

**Review by rhymes — CHANGES_REQUESTED** [review]
Hi @ShubhankarKG, thanks for kickstarting this. I left a bunch of notes inline, in addition we should definitely take care of the following things:

- preview is not working, as this is a script and not an iframe, the script should be ran dynamically, see https://github.com/forem/forem/blob/7a6ad7b83ab223e980479709e69dc35eb4b1f4d6/app/views/articles/show.html.erb#L238-L248 and https://github.com/forem/forem/blob/7a6ad7b83ab223e980479709e69dc35eb4b1f4d6/app/views/comments/index.html.erb#L133-L139 for how that is done (plus the correspondent tags)
- tests are missing
- we need CSS to limit the vertical length of the container div, otherwise it can be very very long, see the screenshot:

Screenshot_2020-11-05 test emgithub

ps. don't forget to prefix the branch with your username the next time, see the [contributing guide](https://docs.forem.com/contributing/forem/)

**Comment by ShubhankarKG** [comment]
Screenshot from 2020-11-13 13-37-27
Screenshot from 2020-11-13 13-37-18

Update after this commit, that is. Will work on the CSS as and tests as well

**Comment by ShubhankarKG** [comment]
If there is a view more button, what should it show? View on GitHub? because running the script will load the entire set if lines aren't provided or the line diff set is huge.

**Comment by ShubhankarKG** [comment]
I'm having some difficulty in setting up this max height thing. Can someone give some pointers please :sweat_smile:? Perhaps a PR that also faced this difficulty would also work. Thanks :pray:

**Comment by msarit** [comment]
Hi @ShubhankarKG! 👋🏾 

I'm returning this PR to draft status, since certain aspects are still being worked out. Please feel free to un-draft it once it is ready! Thanks! 🙏🏾

File: app/javascript/article-form/articleForm.jsx [hunk 3]
```diff
@@ -43,6 +43,16 @@ export default class ArticleForm extends Component {
     }
   }
 
+  static handleEmgithubPreview() {
```
Comment (nickytonline) [inline app/javascript/article-form/articleForm.jsx:46]: refactor (blocking): The article form's purpose it to edit a post. It has no knowledge of liquid tags. Take a peek at existing liquid tags, like this one https://github.com/forem/forem/blob/d9011f4b593469d60499900660818e8ee897c382/app/liquid_tags/user_subscription_tag.rb

**Review by nickytonline — CHANGES_REQUESTED** [review]
@ShubhankarKG, just doing a first pass even though this is in draft.

**Comment by ShubhankarKG** [comment]
Hey everyone, sorry to be saying this but at this moment, I won't be able to continue on this PR :disappointed: I apologize, but I have serious time constraints and so, cannot complete this. I appreciate the help! But again, I apologize.

**Comment by ShubhankarKG** [comment]
Thanks everyone! I look forward to contributing some time in the future!

**Comment by nickytonline** [comment]
No problem @ShubhankarKG. Looking forward to your future contributions!

**Pull request closed without merging**